### ISSUE TITLE
cuda.is_available() is false

### ISSUE THREAD

**Issue by AgentScrubbles** [aoirint/sd-scripts-docker#28]
Hello,

I'm trying to run sdxl training, and the base docker image here is telling me that cuda is not available.  I've tested and can confirm that other containers have cuda enabled and can perform workloads, but only this combination of torch/cuda/whatever is reporting that cuda is not available. 

Full trace:
```
Traceback (most recent call last):
  File "/code/sd-scripts/sdxl_train_network.py", line 183, in <module>
    trainer.train(args)
  File "/code/sd-scripts/train_network.py", line 232, in train
    vae.set_use_memory_efficient_attention_xformers(args.xformers)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 251, in set_use_memory_efficient_attention_xformers
    fn_recursive_set_mem_eff(module)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 247, in fn_recursive_set_mem_eff
    fn_recursive_set_mem_eff(child)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 247, in fn_recursive_set_mem_eff
    fn_recursive_set_mem_eff(child)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 247, in fn_recursive_set_mem_eff
    fn_recursive_set_mem_eff(child)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/modeling_utils.py", line 244, in fn_recursive_set_mem_eff
    module.set_use_memory_efficient_attention_xformers(valid, attention_op)
  File "/home/user/.local/lib/python3.10/site-packages/diffusers/models/attention_processor.py", line 203, in set_use_memory_efficient_attention_xformers
    raise ValueError(
ValueError: torch.cuda.is_available() should be True but is False. xformers' memory efficient attention is only available for GPU 
Traceback (most recent call last):
  File "/home/user/.local/bin/accelerate", line 8, in <module>
    sys.exit(main())
  File "/home/user/.local/lib/python3.10/site-packages/accelerate/commands/accelerate_cli.py", line 47, in main
    args.func(args)
  File "/home/user/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 986, in launch_command
    simple_launcher(args)
  File "/home/user/.local/lib/python3.10/site-packages/accelerate/commands/launch.py", line 628, in simple_launcher
    raise subprocess.CalledProcessError(returncode=process.returncode, cmd=cmd)
subprocess.CalledProcessError: Command '['/opt/python/bin/python3.10', 'sdxl_train_network.py', <<full accelerate command>>
```

nvidia-smi from the host:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 525.125.06   Driver Version: 525.125.06   CUDA Version: 12.0     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  On   | 00000000:01:00.0 Off |                  N/A |
|  0%   53C    P8    15W / 350W |   2019MiB / 24576MiB |      0%      Default |
|                               |                      |                  N/A |
```
(I do note that it's cuda v12, maybe that causes the issue?)

**Comment by Lilneo786** [comment]
GPU Availability: Make sure that your Docker container has access to the GPU. When you launch a Docker container that requires GPU support, you need to specify the --gpus option to grant access to the GPU devices. For example:
`docker run --gpus all -it my_container_image`

**Comment by AgentScrubbles** [comment]
It does, I can run the Nvidia pytorch images just fine, running `python -c import torch; torch.cuda.is_available()` returns true inside those pre built containers - but the docker image here returns false.  The only thing I can think of is some version mismatch somewhere

**Comment by Lilneo786** [comment]
If your Docker container based on an NVIDIA PyTorch image returns False when checking torch.cuda.is_available(), it is possible that there is a version or configuration mismatch issue.

1. Check NVIDIA Driver and CUDA Toolkit Version:
Ensure that the NVIDIA driver on your host system is compatible with the CUDA version used by the PyTorch image in your Docker container. Also, make sure that the versions of PyTorch and CUDA in your Docker image match your host system's capabilities.

2. Verify NVIDIA Container Toolkit:
Confirm that you have the NVIDIA Container Toolkit installed on your host system. This toolkit helps manage GPU resources for Docker containers.

3. Run Container with GPU Access:
When running your Docker container, make sure to use the --gpus all flag to grant GPU access to the container.

4. Check Docker Version:
Ensure that you are using a recent version of Docker that is compatible with NVIDIA GPU support. Upgrade Docker if necessary.

5. Review Dockerfile:
Check the Dockerfile used to create your custom Docker image. Ensure that it correctly installs the necessary GPU drivers, CUDA, cuDNN, and PyTorch versions.

6. Container Environment:
Verify that there are no environment variables or settings within your Docker container that override GPU access.

7. Debug Inside the Container:
If you have access to the container, you can run debugging commands inside it to inspect the GPU environment. Use commands like nvidia-smi and nvcc --version to check the NVIDIA driver and CUDA version.

8. Docker Compose:
If you are using Docker Compose, ensure that your GPU configurations and environment settings are correctly specified in the Compose file.

9. Update PyTorch:
If you suspect a version mismatch, try updating PyTorch to the latest compatible version and rebuild your Docker image.

10. Verify Other Dependencies:
Ensure that all other dependencies in your Docker container are correctly configured for GPU support.

11. Debugging:
Use Python's try-except or other error-handling mechanisms to capture any error messages or exceptions that may occur.

**Comment by aoirint** [comment]
If you are using the `aoirint/sd-scripts` image without modification, please let me know the tag and the command to run (`docker run ...`). Also, any other detailed information on the execution environment might help.

I checked the return value of `torch.cuda.is_available()` in the following tag on my Ubuntu 22.04 environment, but it was `True`, so I could not reproduce the problem.

- (Recommended) `aoirint/sd_scripts:20231108.3@sha256:a370bdf0269da5ff754939203df1e94831ca15fd40e9ab70f34292b51fdf5da2`
- `aoirint/sd_scripts:latest@sha256:cbc4c5172af23b580c5f2aed94f1051e7c976b070cdbd0783263992534768202`

```shell
$ sudo docker run --gpus all --rm \
  --entrypoint "" \
  aoirint/sd_scripts:20231108.3@sha256:a370bdf0269da5ff754939203df1e94831ca15fd40e9ab70f34292b51fdf5da2 \
  gosu user python -c "import torch; print(torch.cuda.is_available())"
True

$ sudo docker run --gpus all --rm \
  --entrypoint "" \
  aoirint/sd_scripts:latest@sha256:cbc4c5172af23b580c5f2aed94f1051e7c976b070cdbd0783263992534768202 \
  gosu user python -c "import torch; print(torch.cuda.is_available())"
True
```

If you are using some forked Docker image, there is not much I can do to help, but you may want to check the following points. The host-side CUDA libraries are not used and are therefore not relevant.

- The CUDA version in the container should be 11.8 (same as this repository)

```shell
$ sudo docker run --gpus all --rm --entrypoint "" aoirint/sd_scripts:20231108.3@sha256:a370bdf0269da5ff754939203df1e94831ca15fd40e9ab70f34292b51fdf5da2 dpkg -l | grep cud
ii  cuda-compat-11-8                520.61.05-1                             amd64        CUDA Compatibility Platform
ii  cuda-cudart-11-8                11.8.89-1                               amd64        CUDA Runtime native Libraries
ii  cuda-keyring                    1.0-1                                   all          GPG keyring for the CUDA repository
ii  cuda-libraries-11-8             11.8.0-1                                amd64        CUDA Libraries 11.8 meta-package
ii  cuda-nvrtc-11-8                 11.8.89-1                               amd64        NVRTC native runtime libraries
ii  cuda-nvtx-11-8                  11.8.86-1                               amd64        NVIDIA Tools Extension
ii  cuda-toolkit-11-8-config-common 11.8.89-1                               all          Common config package for CUDA Toolkit 11.8.
ii  cuda-toolkit-11-config-common   11.8.89-1                               all          Common config package for CUDA Toolkit 11.
ii  cuda-toolkit-config-common      12.1.105-1                              all          Common config package for CUDA Toolkit.
hi  libcudnn8                       8.9.0.131-1+cuda11.8                    amd64        cuDNN runtime libraries
hi  libnccl2                        2.15.5-1+cuda11.8                       amd64        NVIDIA Collective Communication Library (NCCL) Runtime

$ sudo docker run --gpus all --rm --entrypoint "" aoirint/sd_scripts:latest@sha256:cbc4c5172af23b580c5f2aed94f1051e7c976b070cdbd0783263992534768202 dpkg -l | grep cud
ii  cuda-compat-11-8                520.61.05-1                             amd64        CUDA Compatibility Platform
ii  cuda-cudart-11-8                11.8.89-1                               amd64        CUDA Runtime native Libraries
ii  cuda-keyring                    1.0-1                                   all          GPG keyring for the CUDA repository
ii  cuda-libraries-11-8             11.8.0-1                                amd64        CUDA Libraries 11.8 meta-package
ii  cuda-nvrtc-11-8                 11.8.89-1                               amd64        NVRTC native runtime libraries
ii  cuda-nvtx-11-8                  11.8.86-1                               amd64        NVIDIA Tools Extension
ii  cuda-toolkit-11-8-config-common 11.8.89-1                               all          Common config package for CUDA Toolkit 11.8.
ii  cuda-toolkit-11-config-common   11.8.89-1                               all          Common config package for CUDA Toolkit 11.
ii  cuda-toolkit-config-common      12.1.105-1                              all          Common config package for CUDA Toolkit.
hi  libcudnn8                       8.9.0.131-1+cuda11.8                    amd64        cuDNN runtime libraries
hi  libnccl2                        2.15.5-1+cuda11.8                       amd64        NVIDIA Collective Communication Library (NCCL) Runtime
```

- The PyTorch in the container must support CUDA 11.8 (or the CUDA version shown above)

```shell
$ sudo docker run --gpus all --rm \
  --entrypoint "" \
  aoirint/sd_scripts:20231108.3@sha256:a370bdf0269da5ff754939203df1e94831ca15fd40e9ab70f34292b51fdf5da2 \
  gosu user python -c "import torch; print(torch.version.cuda)"
11.8

$ sudo docker run --gpus all --rm \
  --entrypoint "" \
  aoirint/sd_scripts:latest@sha256:cbc4c5172af23b580c5f2aed94f1051e7c976b070cdbd0783263992534768202 \
  gosu user python -c "import torch; print(torch.version.cuda)"
11.8
```

**Comment by AgentScrubbles** [comment]
Thank you @aoirint for all of your great debugging steps!  it helped out a lot.

I was able to find out that my GPU was being passed in, but for underlying reasons CUDA was not.  I'm running kubernetes and from what I was trying it just was not adding CUDA in.

I was able to debug that more in depth using your items here, and then that helped me solve the underlying problem.  (I didn't have a runtime class specified.  (I guess in Docker terms I was passing in my GPUs but the runtime environment wasn't being hooked in, so I was seeing a GPU, but couldn't run any compute loads.)

Thank you for your debugging steps, closing the issue as unrelated.
